stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/utils/RetryUtils.kt

```diff
@@ -0,0 +1,44 @@
+package io.getstream.chat.android.livedata.utils
+
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.call.Call
+import io.getstream.chat.android.client.logger.TaggedLogger
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.livedata.extensions.isPermanent
+import kotlinx.coroutines.delay
+
+internal object RetryUtils {
+
+    internal suspend fun <T : Any> runAndRetry(
```

Comment:
Why do we need to duplicate this from `ChatDomainImpl`?

Reply:
I use a mock for the `ChatDomainImpl` in the unit test... when the  `runAndRetry` gets called, it always returns `null`, so it is not possible to unit test it.

Reply:
You can mock `runAndRetry` with something like this if you need to:

```kotlin
runBlocking {
    whenever(it.runAndRetry<Message>(any())) doAnswer {
        (it.arguments[0] as () -> Call<Message>).invoke().execute()
    }
}
```

Also, we have a lot of setup code in that test that creates a client, domain, etc, perhaps it should extend `BaseDomainTest2` which already does all of that?

Reply:
The mock makes sense. But I would prefer no to use `BaseDomainTest2` it add much more complexity than I would like... hieritance is normally more annoying than helpful.

Reply:
Our unit test setups are pretty complex, so it helps if they at least work the same way. But if you don't want the base test, using things like `TestCoroutineRule` as it's used in the `BaseDomainTest2` might still help simplify your test.